src/analysis.zig

```diff
@@ -2639,19 +2640,62 @@ fn resolveTypeOfNodeUncached(analyser: *Analyser, node_handle: NodeWithHandle) e
         .mod,
         .mul_wrap,
         .mul_sat,
-        .add,
-        .sub,
         .add_wrap,
         .sub_wrap,
         .add_sat,
         .sub_sat,
-        .shl,
-        .shl_sat,
-        .shr,
         .bit_and,
         .bit_xor,
         .bit_or,
-        .bit_not,
+        => {
+            const lhs, const rhs = tree.nodeData(node).node_and_node;
+            const lhs_ty = try analyser.resolveTypeOfNodeInternal(.of(lhs, handle)) orelse return null;
+            const rhs_ty = try analyser.resolveTypeOfNodeInternal(.of(rhs, handle)) orelse return null;
+            if (lhs_ty.is_type_val) return null;
+            if (rhs_ty.is_type_val) return null;
```

Comment:
No need to resolve the rhs when we are going to discard the result anyway.
```suggestion
            const lhs_ty = try analyser.resolveTypeOfNodeInternal(.of(lhs, handle)) orelse return null;
            if (lhs_ty.is_type_val) return null;
            const rhs_ty = try analyser.resolveTypeOfNodeInternal(.of(rhs, handle)) orelse return null;
            if (rhs_ty.is_type_val) return null;
```

The same change should be applied to the other cases.